### ISSUE TITLE
Design API for @esri/arcgis-rest-feature-layers module

### ISSUE THREAD

**Issue by patrickarlt** [Esri/arcgis-rest-js#46]
API design for a module with support for creating, updating and deleting feature both feature layers and their features.

**Comment by tomwayson** [comment]
finally working on this. - I'm thinking of going w/ `arcgis-rest-features` what say ye?

**Comment by tomwayson** [comment]
I was thinking that would have been a nice terse name for a module that works w/ features (query, CRUD, etc). However, I wasn't thinking about the `getServiceInfo()`, `getLayerInfo()`, etc, which I think should be grouped w/ the query/CRUD fns, but seem a bit out of place under that name.

So I'm going to go back to something more along the lines of @noahmulfinger's proposal, but just make it: `arcgis-rest-feature-service` (drop the `s`).

**Comment by tomwayson** [comment]
@noahmulfinger @jgravois 

After #126 the last couple functions I want to add before the initial release of a "read-only" `@esri/arcgis-rest-feature-service` are the functions to get the feature service info and (i.e. `/?f=json`) and the layer info (i.e. `/0?f=json`).

I think we should figure out the naming convention for these types of functions.

I've actually already written `getLayerInfo(url, requestOptions)`, and was planning to write `getServiceInfo(url, requestOptions)`, but then I noticed that @noahmulfinger suggested above to use the name `getService(url, requestOptions)`. Not to be confused w/ @jgravois' `serviceInfo(url, requestOptions)`, which I suggest we rename to include the term "geocode" somewhere (#122). Also I suspect there will be more of these types of functions as we support more types of services, so maybe we need to be more specific than `getService()` or `getServiceInfo()` (i.e. `getFeatureServiceInfo()`). Beyond that, it seems like the general naming convention questions are:
- include `Info` suffix?
- prefix w/ `get`?

Which brings me to my next question - do we even need these functions? I know in the case of @jgravois' `getService()` it defaults to the world geocoder if no URL is passed, so that definitely has a reason to exist. However, the only things my `getLayerInfo()` does beyond `request()` is: 
- default to `GET` (which implicitly depends on #127)
- defines a `ILayerInfo` for the response

I got about 30 lines into that `ILayerInfo` definition before I began to question my life choices:

```ts
export interface ILayerInfo {
  currentVersion: number;
  id: number;
  name: string;
  type: string;
  // TODO: how much of this is optional?
  serviceItemId?: string;
  displayField?: string;
  description?: string;
  copyrightText?: string;
  defaultVisibility?: boolean;
  isDataVersioned?: boolean;
  supportsAppend?: boolean;
  supportsCalculate?: boolean;
  supportsTruncate?: boolean;
  supportsAttachmentsByUploadId?: boolean;
  supportsAttachmentsResizing?: boolean;
  supportsRollbackOnFailureParameter?: boolean;
  supportsStatistics?: boolean;
  supportsAdvancedQueries?: boolean;
  supportsValidateSql?: boolean;
  supportsCoordinatesQuantization?: boolean;
  supportsApplyEditsWithGlobalIds?: boolean;
  advancedQueryCapabilities?: {
    supportsPagination?: boolean;
    supportsPaginationOnAggregatedQueries?: boolean;
    supportsQueryRelatedPagination?: boolean;
    supportsQueryWithDistance?: boolean;
    supportsReturningQueryExtent?: boolean;
    supportsStatistics?: boolean;
    supportsOrderBy?: boolean;
    supportsDistinct?: boolean;
    supportsQueryWithResultType?: boolean;
    supportsSqlExpression?: boolean;
    supportsAdvancedQueryRelated?: boolean;
    supportsCountDistinct?: boolean;
    supportsReturningGeometryCentroid?: boolean;
    supportsQueryWithDatumTransformation?: boolean;
    supportsHavingClause?: boolean;
    supportsOutFieldSQLExpression?: boolean;
  };
  useStandardizedQueries?: boolean;
  geometryType?: esriGeometryType;
  minScale?: number;
  maxScale?: number;
  extent?: IExtent;
  allowGeometryUpdates?: boolean;
  hasAttachments?: boolean;
  // TODO: htmlPopupType type?
  htmlPopupType?: string;
  hasM?: boolean;
  hasZ?: boolean;
  objectIdField?: string;
  globalIdField?: string;
  typeIdField?: string;
  fields?: [IField];
  supportedQueryFormats?: string;
  hasStaticData?: boolean;
  maxRecordCount?: number;
  standardMaxRecordCount?: number;
  tileMaxRecordCount?: number;
  maxRecordCountFactor?: number;
  capabilities?: string;
  [key: string]: any;
  // there is so much more crazy crap that could be added
}
```

So, to recap, I do think we be writing these functions even if they are just very thin wrappers around `request()`, do you agree?

Should we continue to default to `GET`? If so, that makes me lean towards prefixing the function names w/ `get`.

Do we suffix w/ `Info`? I'm leaning towards not.

Do we define interfaces for the response? If so base the name (i.e. `ILayerInfo` or `ILayer`) on what we decided for the function name suffix.

**Comment by tomwayson** [comment]
For consistency w/ #134 I suggest: 

```ts
// both default to GET
getFeatureService(url, requestOptions): IFeatureService;
getFeatureLayer(url, requestOptions): IFeatureLayer;
```

**Comment by noahmulfinger** [comment]
@tomwayson The way you set up the functions listed at the bottom looks good to me. Defaulting to GET makes sense, and having an `IFeatureService` interface matches up with the function names.  I don't have a strong opinion on whether to add `Info`.

In the admin package, I'm also adding 
```ts
getFeatureServiceAdmin(url, requestOptions): IFeatureServiceAdmin;
```

Not super happy with just tacking on Admin, but can't think of a better way.

**Comment by tomwayson** [comment]
FYI - that's what we do in Hub, we have several cases of a `fooBar` service w/ an associated  `fooBarAdmin`. The semantics feel a bit awkward to me, but the responsibilities of each service are totally clear. So that works for me.
